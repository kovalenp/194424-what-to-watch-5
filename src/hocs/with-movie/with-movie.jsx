import React from "react";
import PropTypes from "prop-types";
import {connect} from "react-redux";

import {getCurrentMovie} from "../../services/movie-service";

const withMovie = (RenderComponent) => {
  function WithMovie(props) {

    if (!props.movie || props.movie.id !== parseInt(props.id, 10)) {
      const {getMovie} = props;
      getMovie();
      return null;
    }

    return (
      <RenderComponent
        {...props}
        movie={props.movie}
      />
    );

  }

  WithMovie.propTypes = {
    id: PropTypes.string.isRequired,
    getMovie: PropTypes.func,
    movie: PropTypes.object,
    movies: PropTypes.array,
  };

  const mapStateToProps = (state) => {
    return {
      movie: state.MOVIES.current,
      movies: state.MOVIES.list,
    };
  };

  const mapDispatchToProps = (dispatch, ownProps) => {
    return {
      getMovie: () => dispatch(getCurrentMovie(ownProps.id))
    };
  };

  return connect(mapStateToProps, mapDispatchToProps)(WithMovie);
};

export {withMovie};
export default withMovie;
