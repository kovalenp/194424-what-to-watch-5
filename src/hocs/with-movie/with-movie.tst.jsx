import React from "react";
import renderer from "react-test-renderer";
import PropTypes from "prop-types";
import {withMovie} from "./with-movie";
import configureMockStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import {Provider} from "react-redux";
import movie from "../../mocks/movie";

const middlewares = [thunk];
const mockStore = configureMockStore(middlewares);
const store = mockStore({MOVIES: {current: movie, list: [movie]}});

const MockComponent = (props) => {
  const {children} = props;

  return (
    <div>
      {children}
    </div>
  );
};

MockComponent.propTypes = {
  children: PropTypes.oneOfType([
    PropTypes.arrayOf(PropTypes.node),
    PropTypes.node
  ]).isRequired,
};

const MockComponentWrapped = withMovie(MockComponent);

describe(`withMovie meets B21`, () => {

  beforeAll(() => {
    jest.mock(`../../services/movie-service`);
  });

  afterAll(() => {
    jest.mockClear();
  });

  it(`and rendered correctly (nah, just kidding)`, () => {
    const tree = renderer.create((
      <Provider store={store}>
        <MockComponentWrapped>
          <React.Fragment />
        </MockComponentWrapped>
      </Provider>
    )).toJSON();

    expect(tree).toMatchSnapshot();
  });
});
