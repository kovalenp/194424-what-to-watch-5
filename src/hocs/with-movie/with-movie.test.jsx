import React from "react";
import renderer from "react-test-renderer";
import PropTypes from "prop-types";
import {withMovie} from "./with-movie";
import configureMockStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import {Provider} from "react-redux";
import movie from "../../mocks/movie";

const middlewares = [thunk];
const mockStore = configureMockStore(middlewares);
const store = mockStore({MOVIES: {current: movie, list: [movie]}});

jest.mock(`../../services/movie-service`, () => ({
  __esModule: true,
  default: () => () => <p>MockHeader</p>,
  getCurrentMovie: () => ({type: `MOCK`})
}));

const MockComponent = (props) => {
  const {children} = props;

  return (
    <div>
      {children}
    </div>
  );
};

MockComponent.propTypes = {
  children: PropTypes.oneOfType([
    PropTypes.arrayOf(PropTypes.node),
    PropTypes.node
  ]).isRequired,
};

const MockComponentWrapped = withMovie(MockComponent);

describe(`withMovie meets B21`, () => {

  afterAll(() => {
    jest.clearAllMocks();
  });

  it(`and rendered correctly (nah, just kidding)`, () => {
    const tree = renderer.create((
      <Provider store={store}>
        <MockComponentWrapped>
          <React.Fragment />
        </MockComponentWrapped>
      </Provider>
    )).toJSON();

    expect(tree).toMatchSnapshot();
  });
});
