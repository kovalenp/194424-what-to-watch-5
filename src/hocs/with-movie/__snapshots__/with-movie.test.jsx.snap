// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`withMovie meets B21 and rendered correctly (nah, just kidding) 1`] = `null`;
